be/src/service/service_be/lake_service.cpp

```diff
@@ -243,18 +243,25 @@ void LakeServiceImpl::publish_version(::google::protobuf::RpcController* control
                     if (res.ok()) {
                         auto metadata = std::move(res).value();
                         auto score = compaction_score(_tablet_mgr, metadata);
-                        std::lock_guard l(response_mtx);
-                        response->mutable_compaction_scores()->insert({tablet_id, score});
-                        if (request->base_version() == 1) {
-                            int64_t row_nums = std::accumulate(
-                                    metadata->rowsets().begin(), metadata->rowsets().end(), 0,
-                                    [](int64_t sum, const auto& rowset) { return sum + rowset.num_rows(); });
-                            // Used to collect statistics when the partition is first imported
-                            response->mutable_tablet_row_nums()->insert({tablet_id, row_nums});
+                        TabletMetadataPB* prealloc_metadata = nullptr;
+                        {
+                            std::lock_guard l(response_mtx);
+                            response->mutable_compaction_scores()->insert({tablet_id, score});
+                            if (request->base_version() == 1) {
+                                int64_t row_nums = std::accumulate(
+                                        metadata->rowsets().begin(), metadata->rowsets().end(), 0,
+                                        [](int64_t sum, const auto& rowset) { return sum + rowset.num_rows(); });
+                                // Used to collect statistics when the partition is first imported
+                                response->mutable_tablet_row_nums()->insert({tablet_id, row_nums});
+                            }
+                            if (skip_write_tablet_metadata) {
+                                auto& map = *response->mutable_tablet_metas();
+                                prealloc_metadata = &map[tablet_id];
```

Comment:
TabletMetadataPB is pretty large, this `map[tablet_id]` triggers the allocation and initialization of the PB instance, which will be still costly. So not quite sure how much will be improved by allocating here and then CopyFrom outside the lock.

Reply:
ok